# 課題2 データの集計

## Q1
中央値はデータの中心傾向を観察するための記述統計の指標です。同じくデータの中心傾向の指標である平均値と比べると、データの外れ値の影響を受けにくという特徴があります。

1変数の$n$個のデータを考えた時、中央値を求めるにはまず、データの値の大きさの順に並び替えます。次に、データの数が奇数個の場合は、大きさの順に並び替えた$(n+1)/2$番目の値、データの数が偶数個の場合は、大きさの順に並び替えた$n/2$番目と$n/2+1$の値の平均の値、をそれぞれデータの中央値とします。

分散はデータのばらつきを観察するための記述統計の指標です。個々のデータがデータ全体の平均からどらぐらいばらついているかを表します。以下の通り、個々のデータの値$x_{i}$とデータ全体の平均の値$\overline{x}$の差の2乗和をデータ数$n$で割ったものが分散になります。

$\frac{1}{n}\Sigma^{n}_{i=1}(x_{i}-\overline{x})^{2}$

0から100までの整数を要素とする任意の長さのリストを引数で入力として受け取り、以下を要素とするリストを返す`avg_med_var()`関数を完成させてください。
- 入力リストの全ての要素の平均値
- 入力リストの全ての要素の中央値
- 入力リストの全ての要素の分散

In [1]:
def avg_med_var(input_list):
    mean_value=0 #平均値を初期化
    med_value=0 # 中央値を初期化
    var_value=0 # 分散を初期化
    results=[] # 結果のリスト
        
    input_list.sort() # リストの要素を昇順に並び替え
    mid_idx = len(input_list)//2 # リストの中央（長さの半分）のインデックス（必要であれば利用する）
    
    if len(input_list)%2 ==0:### リストの長さが偶数か奇数か調べる条件のコード ###
        med_value= (input_list[mid_idx]+input_list[mid_idx+1])/2### リストの長さが偶数の時の中央値を計算するコード ###
    else:
        med_value = input_list[mid_idx]### リストの長さが奇数の時の中央値を計算するコード ###
    
    mean_value = sum(input_list)/len(input_list) #平均値
    mse_value = 0 # データの値と平均値の差の2乗和を初期化
    
    for lis in input_list:### input_listの要素を1つずつ処理するコード ###
        mse_value += (lis - mean_value) ** 2 ### 要素の数だけデータの値と平均値の差の2乗和を更新するコード ###
    var_value = mse_value / len(input_list) ### 分散を計算するコード ###
    
    results.append(mean_value) 
    results.append(med_value) 
    results.append(var_value)
    return results   

`avg_med_var()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [2]:
print(avg_med_var([1,4,3,2,5,6,8,7,9,10]))
print(avg_med_var([1,4,3,2,5,6,8,7,9,10,11]))

[5.5, 6.5, 8.25]
[6.0, 6, 10.0]


以下のセルを実行して、実行結果が上記と同じであること確認してください。結果が異なる場合は、作成した`avg_mean_var`関数に誤りがないか確認してください。

In [3]:
# avg_med_var関数の結果を検証するための関数
import numpy as np
def avg_med_var_test(input_list):
    return [np.mean(input_list), np.median(input_list), np.var(input_list)]

In [4]:
print(avg_med_var_test([1,4,3,2,5,6,8,7,9,10]))
print(avg_med_var_test([1,4,3,2,5,6,8,7,9,10,11]))

[5.5, 5.5, 8.25]
[6.0, 6.0, 10.0]


# Q2
以下のような形式の"user_score.csv"ファイルを読み込み、データフレーム`score`を作成します。
```Python
## user_score.csvファイル
user, kokugo, shakai, sugaku, rika
1, 30, 43, 51, 63
2, 39, 21, 49, 56
...
```

In [5]:
import pandas as pd
score = pd.read_csv('user_score.csv')
score.head(5)

user  kokugo  shakai  sugaku  rika
0     1      30      43      51    63
1     2      39      21      49    56
2     3      29      30      23    57
3     4      95      87      77   100
4     5      70      71      78    67

データフレームの`desribe()`メソッドを使って、各教科ごとの記述統計を確認してください

In [6]:
score.describe()

user      kokugo      shakai      sugaku        rika
count  166.000000  166.000000  166.000000  166.000000  166.000000
mean    83.500000   52.331325   39.602410   45.614458   49.885542
std     48.064193   21.835858   21.456335   24.236127   21.408130
min      1.000000    0.000000    4.000000    0.000000    9.000000
25%     42.250000   35.000000   23.000000   28.000000   31.000000
50%     83.500000   53.000000   35.000000   43.000000   49.000000
75%    124.750000   70.000000   53.750000   63.750000   65.750000
max    166.000000   96.000000   90.000000  100.000000  100.000000

### Q2-1
データフレーム`score`から各教科の得点がすべて平均点以上であるようなユーザの数を返す`above_mean()`関数を作成してください。以下のようにして、データフレームの指定した列の平均値を求めることができます。
```Python
データフレーム[列].mean()
```

In [7]:
'''
def above_mean(df):
    return len(df[### 各教科の得点がすべて平均点以上という条件のコード ###])
'''

#df[(df['kokugo'] > 50) & (df['shakai'] > 50)]
def above_mean(df):
    return len(df[(df['kokugo'] > df['kokugo'].mean()) & (df['shakai'] > df['shakai'].mean()) & (df['sugaku'] > df['sugaku'].mean()) & (df['rika'] > df['rika'].mean())])

`above_mean()`関数が完成したら、以下のセルを実行して動作を確認してください。条件満たすユーザは46人となります。

In [8]:
above_mean(score)

46

### Q2-2
ユーザごとの全教科の得点の総和の値を表す'sum'列をデータフレーム`score`に新たに追加したデータフレームを返す`score_sum()`関数を作成してください。以下のようにして、データフレームの指定した列の行方向の値の和を持つシリーズオブジェクト作ることができます。
```Python
データフレーム[[列1,列2,...,]].sum(axis=1)
```

In [9]:
def score_sum(df):
    df['sum']= df[['kokugo','shakai','sugaku','rika']].sum(axis=1)### ユーザごとの全教科の得点の総和の値を持つシリーズオブジェクトを作成するコード ###
    return df

`score_sum()`関数が完成したら、以下のセルを実行して動作を確認してください。'sum'列が追加されていることを確認してください。

In [10]:
score = score_sum(score)
score.head(5)

user  kokugo  shakai  sugaku  rika  sum
0     1      30      43      51    63  187
1     2      39      21      49    56  165
2     3      29      30      23    57  139
3     4      95      87      77   100  359
4     5      70      71      78    67  286

In [11]:
 score['sum'].iloc[0]

187

全教科の合計得点の上位3人の'user'を1位、2位、3位の順に要素として持つリストを返す`score_top3()`関数を作成させてください。

In [12]:
def score_top3(df):
    sorted_df = df.sort_values(by='sum', ascending=False)### データフレームを'sum'列の値で並び替えて新しいデータフレームを作成するコード ###
    results=[]
    results.append(sorted_df['user'].iloc[0])### 'sum'が第1位の'user'を抽出するコード ###
    results.append(sorted_df['user'].iloc[1])### 'sum'が第2位の'user'を抽出するコード ###
    results.append(sorted_df['user'].iloc[2])### 'sum'が第3位の'user'を抽出するコード ###
    return results

`score_top3()`関数が完成したら、以下のセルを実行して動作を確認してください。`[39, 4, 50]`の'user'が合計得点の上位3人です。

In [13]:
score_top3(score)

[39, 4, 50]

### Q2-3
データフレーム`score`の'sum'列を対象に、以下を階級幅として各階級の度数を要素として持つリストを返すscore_hist()関数を完成させてください。
- 0以上、50以下
- 51以上、100以下
- 101以上、150以下
- 151以上、200以下
- 201以上、250以下
- 251以上、300以下
- 301以上、350以下
- 351以上、400以下

In [14]:
def score_hist(df):
    results=[]
    results.append(((0<=df['sum']) & (df['sum']<=50)).sum())### 'sum'が0以上、50以下のデータ数を抽出するコード ###
    results.append(((51<=df['sum']) & (df['sum']<=100)).sum())### 'sum'が51以上、100以下のデータ数を抽出するコード ###
    results.append(((101<=df['sum']) & (df['sum']<=150)).sum())
    results.append(((151<=df['sum']) & (df['sum']<=200)).sum())
    results.append(((201<=df['sum']) & (df['sum']<=250)).sum())
    results.append(((251<=df['sum']) & (df['sum']<=300)).sum())
    results.append(((301<=df['sum']) & (df['sum']<=350)).sum())
    results.append(((351<=df['sum']) & (df['sum']<=400)).sum())### 'sum'が351以上、400以下のデータ数を抽出するコード ###
    return results

`score_hist()`関数が完成したら、以下のセルを実行して動作を確認してください。`[4, 20, 39, 35, 29, 24, 13, 2]`が'sum'の度数分布です。

In [15]:
score_hist(score)

[4, 20, 39, 35, 29, 24, 13, 2]

### Q2-4
以下のような形式の"user_class.csv"ファイルを読み込み、データフレーム`user_class`を作成します。
```Python
## user_class.csvファイル
'user', 'class'
1, 'C'
2, 'C'
3, 'B'
...
```

In [16]:
user_class = pd.read_csv('user_class.csv')
user_class.head(5)

user class
0     1     C
1     2     C
2     3     B
3     4     C
4     5     A

'user'の列の値をキーに、2つのデータフレーム、user_classとscore、を結合して新しいデータフレームを作成した上で、データフレームの`groupby`メソッドを使って、'class'をインデックス、'class'ごとの'sum'の最高値を列として持つシリーズオブジェクトを返す`score_by_class()`関数を完成させてください。

In [17]:
def score_by_class(df1, df2):
    df = df1.set_index('user').join(df2.set_index('user'))### user_classとscoreを結合して新しいデータフレームを作成するコード ###
    return pd.Series(df.groupby('class')['sum'].max())### groupbyメソッドを使って、'class'ごとの'sum'の最高値をシリーズオブジェクトとして抽出するコード ###

`score_by_class()`関数が完成したら、以下のセルを実行して動作を確認してください。`{'A': 361, 'B': 349, 'C': 359}`が'class'ごとの'sum'の最高値です。

In [18]:
score_by_class(score, user_class).to_dict()    

{'A': 361, 'B': 349, 'C': 359}